src/components/Input/index.tsx

```diff
@@ -0,0 +1,162 @@
+import { Transition } from '@headlessui/react';
+import { InformationCircleIcon } from '@heroicons/react/outline';
+import {
+  ExclamationCircleIcon,
+  EyeIcon,
+  EyeOffIcon,
+} from '@heroicons/react/solid';
+import classNames from 'classnames';
+import React, { useEffect, useState, ChangeEventHandler } from 'react';
+
+export interface InputProps {
+  type: 'text' | 'number' | 'password';
+  onChange: ((v: string) => void) | ((v: number) => void);
+  value: string | number;
+  label?: string;
+  description?: string;
+  placeholder?: string;
+  required?: boolean;
+  maxLength?: number;
+  min?: number;
+  max?: number;
+  focusColor?: string;
+  error?: string;
+}
+
+export const Input: React.FC<InputProps> = ({
+  type = 'text',
+  onChange,
+  value,
+  label,
+  description,
+  placeholder,
+  required,
+  maxLength,
+  min,
+  max,
+  focusColor = 'focus:ring-blue-300',
+  error,
+}) => {
+  const [tooltip, setTooltip] = useState('');
+  const [tooltipShow, setTooltipShow] = useState(false);
+  const [passwordVis, setPasswordVis] = useState(false);
+
+  useEffect(() => {
+    if (tooltipShow) {
+      const timeout = setTimeout(() => {
+        setTooltipShow(false);
+      }, 2500);
+      return () => clearTimeout(timeout);
+    }
+    return;
+  }, [tooltipShow]);
+
+  const handleChange: ChangeEventHandler<HTMLInputElement> = (e) => {
+    e.preventDefault();
+    let val = e.target.value;
+    if (type === 'number') {
+      if (min && +val < min) {
+        setTooltip(`Your number must be ${min} or above`);
+        setTooltipShow(true);
+        return;
+      }
+      if (max && +val > max) {
+        setTooltip(`Your number must be ${max} or under`);
+        setTooltipShow(true);
+        return;
+      }
+    } else {
+      if (maxLength && val.length > maxLength) {
+        setTooltip(`Your text can be at most ${maxLength} characters long`);
+        setTooltipShow(true);
+        return;
+      }
+    }
+    onChange((type === 'number' ? +val : val) as never);
+  };
+
+  const handleType = (): string => {
+    if (type === 'password' && passwordVis) return 'text';
+    return type;
+  };
+
+  return (
+    <div>
+      <label
```

Comment:
label is optional

Reply:
If it is undefined then it will act as if the parent did not exist with 0 width and height. This works, tested in the viewing.

Reply:
but wouldn't it leave an unused <label> in DOM?

Reply:
What's so bad about that? It saves on unnecessary logic.